index 22ca4c2..ccd70e7 100644
--- a/atom/atom.ml
+++ b/atom/atom.ml
@@ -22,6 +22,7 @@
 module Xml = Xmlm
 
 type field = [
+  | `XML of string 
   | `Text of string
   | `Empty
 ]
@@ -89,4 +89,6 @@ let output_field t (f:field) =
     | `Text s ->
         Xml.output o (`Data s)
+    | `XML h ->
+        Xml.output o (`Raw h)
     | `Empty -> ()
   )
index 48da954..c056a34 100644
--- a/atom/atom.mli
+++ b/atom/atom.mli
@@ -15,7 +15,7 @@
  *
  *)
 
-type field = [ `Empty | `Text of string ]
+type field = [ `Empty | `Text of string | `XML of string ]
 type author = { name : string; uri : string option; email : string option; }
 type date = int * int * int * int * int
 type meta = {
